### ISSUE TITLE
Add dataset cleaner and grokking-valid data pipeline

### ISSUE THREAD

**Issue by iamdarshg** [iamdarshg/better-ai#49]
## Problem

The current `datasets.yml` is large and ambitious, but raw dataset volume is not enough to support credible claims about complex grokking, delayed generalization, or emergent coding behavior. The corpus mixes general text, code, math, repo-scale tasks, SFT, RLHF, security DPO, and long-context datasets, which is useful for capability building but dangerous for measurement unless the data is cleaned and split correctly.

Specific risks:

- duplicate and near-duplicate examples across pretraining, SFT, RLHF, security DPO, and eval datasets
- benchmark leakage into HumanEval, SWE-bench, RepoBench, TACO/APPS-style evals, and security/code-repair tests
- train/test splits that are random rather than compositional, making memorization look like generalization
- invalid Hugging Face dataset IDs, configs, or splits that only fail at runtime
- inconsistent fields across datasets (`text`, `content`, `code`, `messages`, `chosen/rejected`, etc.)
- unclear dataset weighting and effective sampling rates across curriculum stages
- long-context examples that may exceed practical memory limits or silently truncate away the important part
- preference datasets with missing or malformed `chosen`/`rejected` pairs
- lack of data cards, license notes, and intended skill labels for each dataset
- no evidence that the corpus actually supports grokking-style delayed validation jumps rather than ordinary memorization or leakage

## Goal

Add a proper dataset cleaner/auditor that turns the current dataset list into a validated, deduplicated, leakage-checked, grokking-aware training corpus.

This should not just "clean text". It should make the training data scientifically usable for delayed-generalization experiments and practically usable for staged training.

## Proposed implementation

Add a new module, for example:

```text
better_ai/data/dataset_cleaner.py
better_ai/data/dataset_audit.py
tools/clean_datasets.py
tools/audit_datasets.py
configs/dataset_cleaning.yml
```

The cleaner should support both a tiny CI-safe mode and a full offline audit mode.

### 1. Dataset load validation

- Validate every dataset in `datasets.yml` before training starts.
- Check dataset path, config name, split, streaming support, and required fields.
- Emit a clear error report instead of failing mid-training.
- Support `--dry-run`, `--max-samples`, and `--stage` filters.

### 2. Schema normalization

Normalize all examples into explicit internal schemas:

```python
PretrainingExample:
    text: str
    source_dataset: str
    stage: str
    language: str | None
    skill_tags: list[str]

SFTExample:
    prompt: str
    response: str
    source_dataset: str
    skill_tags: list[str]

PreferenceExample:
    prompt: str
    chosen: str
    rejected: str
    source_dataset: str
    reward_tags: list[str]
```

Do not let the dataloader guess fields ad hoc for every dataset. The cleaner should make the format explicit.

### 3. Exact and near-duplicate detection

- Add exact hash deduplication.
- Add near-duplicate detection using MinHash/SimHash or another scalable method.
- Deduplicate across stages, not only within each dataset.
- Report duplicate clusters and source datasets responsible.
- Preserve one canonical copy and write a provenance record.

### 4. Benchmark leakage checks

Check contamination against held-out/eval datasets, especially:

- HumanEval
- SWE-bench test
- RepoBench eval/test splits
- TACO/APPS-style tasks
- BigCodeBench eval-like tasks
- security/CVE repair evals
- any future grokking synthetic evals

Use exact match, fuzzy match, and code-aware normalization where possible:

- strip comments/docstrings optionally
- normalize whitespace
- normalize variable names for code snippets where feasible
- compare problem statements separately from solutions

### 5. Compositional grokking splits

Generate splits that can actually test delayed generalization:

- short programs train, longer programs validation/test
- small graph sizes train, larger/composed graph sizes test
- single-file bugs train, multi-file bugs test
- shallow nesting train, deeper nesting test
- seen APIs train, unseen API combinations test
- simpler vulnerability patterns train, composed/obfuscated variants test
- train on individual algorithm families, test on compositions of families

This should produce split metadata explaining why each eval item is out-of-distribution relative to train.

### 6. Long-context quality checks

For examples with `max_seq_length` above 32k:

- estimate token length distribution
- detect examples where truncation removes the answer/patch/critical context
- mark examples as repo-scale, document-scale, notebook-scale, or conversation-scale
- report examples that are long but low-signal boilerplate
- optionally chunk long examples into meaningful units with provenance

### 7. Stage and domain weighting report

Emit JSON/CSV reports showing:

- examples per dataset
- estimated tokens per dataset
- effective sampling weight per stage/domain
- mean/median/p95 sequence length
- duplicate rate
- leakage risk score
- malformed example rate
- language distribution
- skill/domain distribution
- recommended upweight/downweight/drop decisions

### 8. Preference-data validation

For RLHF/DPO/security DPO datasets:

- verify `chosen` and `rejected` exist and are not identical
- verify prompt/context is present
- detect empty, ultra-short, or malformed pairs
- detect cases where `chosen` is obviously worse by heuristic checks where possible
- report chosen/rejected length imbalance
- mark datasets with weak preference quality

### 9. Data cards and provenance

Generate a data-card-style summary for each dataset:

- dataset name/path/config/split
- stage
- intended skill
- license/permissiveness note if available
- fields used
- cleaning decisions
- duplicate/leakage stats
- known limitations
- reason for inclusion

### 10. CI-safe validation

Add a tiny CI job that runs:

```bash
python tools/audit_datasets.py --config datasets.yml --max-samples 10 --dry-run
python tools/clean_datasets.py --config datasets.yml --stage sft --max-samples 10 --output /tmp/cleaned
```

This should not require downloading the full corpus.

## Acceptance criteria

- A dataset audit command can validate all entries in `datasets.yml` and produce a structured report.
- Cleaner outputs normalized examples with explicit schemas instead of relying on ad hoc dataloader field guessing.
- Exact and near-duplicate detection works across stages.
- Leakage report is generated against all eval datasets.
- Compositional grokking splits are generated for at least 5 task families.
- Long-context examples are checked for truncation and signal quality.
- Preference datasets are validated for malformed or degenerate chosen/rejected pairs.
- CI includes a tiny dataset audit/cleaner smoke test.
- Documentation explains how to interpret the cleaning report and how to decide whether to add/drop/reweight datasets.

## Why this matters

Without this, the repo can train a large model and still be unable to tell whether it learned reusable abstractions or just memorized leaked/duplicated benchmark-like examples.

With this, delayed validation jumps in coding, repo reasoning, security repair, and tool-use tasks become much more credible evidence of complex grokking/emergent behavior.
